### PR TITLE
Add server log to feedback

### PR DESCRIPTION
**Pull request by stevemessick**
@alexander-doroshko Add most recent server log lines to the feedback file. Make the file always get written.

This may take a few iterations to determine just what log info the analyzer teams wants to see.

cc @devoncarew

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java [hunk 1]
```diff
@@ -140,6 +143,14 @@
   public long maxMillisToWaitForServerResponse = 0L;
   @NotNull private final InteractiveErrorReporter myErrorReporter = new InteractiveErrorReporter();
 
+  @NotNull private final Object myDebugLogLock = new Object();
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: Maybe use the collection itself as a lock, just to save a field?

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java [hunk 2]
```diff
@@ -140,6 +143,14 @@
...
+  @NotNull private final LinkedHashMap<String, String> myDebugLog = new LinkedHashMap<String, String>(DEBUG_LOG_CAPACITY) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: Maybe use some standard solution? I've found `com.google.common.collect.EvictingQueue` that seems to serve the purpose.

File: Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java [hunk 3]
```diff
@@ -29,21 +29,19 @@ public void sendFeedback(@Nullable Project project, @Nullable String errorMessag
...
-          template = potentialTemplate.substring(0, MAX_URL_LENGTH);
-        }
-        catch (IOException e) {
-          // ignore it
+      try {
+        File file = FileUtil.createTempFile("report", ".txt");
+        FileUtil.writeToFile(file, errorMessage);
+        if (serverLog != null) {
+          // Assume serverLog is never long enough that opening and closing the file is cheaper than copying it.
+          FileUtil.writeToFile(file, "\n\n" + serverLog, true);
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java]: I think we can calculate contents first and then write it to a file in one shot.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 1 above)
Comment (stevemessick) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: That caused IntelliJ to signal the field was accessed in both sync and async contexts.

File: Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java (hunk 3 above)
Comment (stevemessick) [inline Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java]: Is that considered "best practice" at JetBrains? That first string can be quite long, which was why I didn't try combining them, but it looks easy enough to change.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 2 above)
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: It was StackOverflow where I found EvictingQueue :). Queue/List works O(1) and its algorithm is very simple whereas LinkedHashMap adds some overhead. Not a big deal for our case, just `put(str, str)` looked a bit unnatural to me.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 1 above)
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: Need to check. If inspection is buggy I'll file an issue.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 2 above)
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: As one more pro I guess `debugLogContent()` implementation would be much simpler with simple collection :)

File: Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java (hunk 3 above)
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DefaultDartFeedbackBuilder.java]: I do not have any metrics in hands to support one or another approach. Just intuitively thought that string concatenation is more efficient than appending a file on a disk. Not a big deal anyway, this code is not called in a loop.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 1 above)
Comment (stevemessick) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: I double-checked this and found that it works as expected. Removing the lock field is fine. I can't explain why it was complaining yesterday. I can change this when I get back to work after the holidays.

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java (hunk 2 above)
Comment (stevemessick) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: I must be missing something. The complexity is due to reversing the order of the entries. Changing to a queue doesn't help with that. I don't think it is worth the trouble to rewrite.

**Comment by alexander-doroshko** [comment]
Ok, pushed as is so far.

**Comment by alexander-doroshko** [comment]
Also caught an exception and pushed a fix
```
java.lang.StringIndexOutOfBoundsException: String index out of range: 4000
	at java.lang.String.substring(String.java:1963)
	at com.jetbrains.lang.dart.ide.errorTreeView.DefaultDartFeedbackBuilder.sendFeedback(DefaultDartFeedbackBuilder.java:41)
```
BTW do we still need to care about URL length since it doesn't include stack any more?

**Pull request merged**